### PR TITLE
Docs: provide Guidance for MultiChildRenderObjectWidget implementations

### PR DESCRIPTION
**Pull request by creativecreatorormaybenot**
## Description

Given how the framework implements `MultiChildRenderObjectWidget` and when the mentioned mixins are used, this should be valuable information for anyone trying to learn about the widget.

## Tests

*Docs only*

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/object.dart [hunk 1]
```diff
@@ -3001,6 +3001,21 @@ mixin ContainerParentDataMixin<ChildType extends RenderObject> on ParentData {
 ///
 /// Provides a child model for a render object subclass that has a doubly-linked
 /// list of children.
+/// 
+/// The [ChildType] specifies the type of render object for the children,
+/// e.g. [RenderBox].
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/object.dart]: nit: separate paragraphs with one empty line.

File: packages/flutter/lib/src/rendering/object.dart [hunk 2]
```diff
@@ -3001,6 +3001,21 @@ mixin ContainerParentDataMixin<ChildType extends RenderObject> on ParentData {
...
+/// For the parent data of the children, the [ParentDataType] is required to 
+/// extend [ContainerParentDataMixin], which provides the necessary functionality to
+/// grab the siblings of [firstChild] or [lastChild]. 
+/// This data will be accessible via [RenderObject.parentData] of the children if 
+/// the children are set up correctly in [RenderObject.setupParentData] of the
+/// parent render object (the one with this mixin).
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/object.dart]: nit: to keep the docs [shorter](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose) I suggest `[ParentDataType] stores parent container data on its child render objects. It must extend [ContainerParentDataMixin], which provides the interface for enumerating children. This data is populated by [RenderObject.setupParentData] implemented by the class using this mixin.`

File: packages/flutter/lib/src/rendering/object.dart [hunk 3]
```diff
@@ -3001,6 +3001,21 @@ mixin ContainerParentDataMixin<ChildType extends RenderObject> on ParentData {
...
+///
+/// If using [RenderBox] for the child type, you will usually want to make use of
+/// [RenderBoxContainerDefaultsMixin] and extend [ContainerBoxParentData] for the
+/// parent data.
+///
+/// Moreover, this is a default mixin for render objects returned to [MultiChildRenderObjectWidget].
```
Comment (yjbanov) [inline packages/flutter/lib/src/rendering/object.dart]: It is more than a default. It is required. `MultiChildRenderObjectElement` will actively cast its render object down to `ContainerRenderObjectMixin`. So you'll get runtime errors if you do not use this mixin.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 4]
```diff
@@ -1799,10 +1799,16 @@ abstract class SingleChildRenderObjectWidget extends RenderObjectWidget {
   SingleChildRenderObjectElement createElement() => SingleChildRenderObjectElement(this);
 }
 
-/// A superclass for RenderObjectWidgets that configure RenderObject subclasses
+/// A superclass for [RenderObjectWidget]s that configure [RenderObject] subclasses
 /// that have a single list of children. (This superclass only provides the
 /// storage for that child list, it doesn't actually provide the updating
 /// logic.)
+///
+/// This will typically return a [RenderBox] mixing in [ContainerRenderObjectMixin]
```
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/framework.dart]: Mixing in `ContainerRenderObjectMixin` is required (see, e.g., `MultiChildRenderObjectElement.insertChildRenderObject`, which unconditionally casts the render object).

File: packages/flutter/lib/src/widgets/framework.dart [hunk 5]
```diff
@@ -1799,10 +1799,16 @@ abstract class SingleChildRenderObjectWidget extends RenderObjectWidget {
...
+/// and [RenderBoxContainerDefaultsMixin], which provide the necessary functionality
+/// to grab the children in the container render object.
+///
+/// See [Stack] and [RenderStack] for an example implementation.
```
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/framework.dart]: nit: we usually format sections like this as follows:

```
/// See also:
///
/// * [Stack], ... insert short explanation of why this example is interesting ...
/// * [RenderStack], ... insert short explanation of why this example is interesting ...
```

File: packages/flutter/lib/src/widgets/framework.dart (hunk 4 above)
Comment (creativecreatorormaybenot) [reply packages/flutter/lib/src/widgets/framework.dart]: The "typically" was referring to the `RenderBox` part. I will see how I can rewrite it.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 6]
```diff
@@ -1799,10 +1799,20 @@ abstract class SingleChildRenderObjectWidget extends RenderObjectWidget {
...
 
-/// A superclass for RenderObjectWidgets that configure RenderObject subclasses
+/// A superclass for [RenderObjectWidget]s that configure [RenderObject] subclasses
 /// that have a single list of children. (This superclass only provides the
 /// storage for that child list, it doesn't actually provide the updating
 /// logic.)
+///
+/// This will return a [RenderObject] mixing in [ContainerRenderObjectMixin].
+/// Typically, this is a [RenderBox] with [RenderBoxContainerDefaultsMixin],
+/// which provide the necessary functionality to grab the children of the
```
Comment (yjbanov) [inline packages/flutter/lib/src/widgets/framework.dart]: nit: let's use "visit" or "traverse" instead of "grab".

**Review by yjbanov — COMMENTED** [review]
One last nit. Thanks for improving our docs!

**Comment by creativecreatorormaybenot** [comment]
@yjbanov Thank you for reviewing the PR :)

**Review by yjbanov — APPROVED** [review]
LGTM. Thanks!

**Comment by yjbanov** [comment]
Landing on red. These are documentation only, and it would be good to kick the bots. Looks like they are just being flaky.

**Pull request merged**